ros/src/computing/perception/localization/packages/kalman_filter_localizer/src/kalman_filter_localizer.cpp

```diff
@@ -0,0 +1,535 @@
+    return;
+  }
+
+  /* Set measurement matrix */
+  Eigen::MatrixXd C = Eigen::MatrixXd::Zero(dim_y, dim_x_);
+  C(0, IDX::VX) = 1.0; // for vx
+  C(1, IDX::WZ) = 1.0; // for wz
+
+  /* Set measurement noise covariancs */
+  Eigen::MatrixXd R = Eigen::MatrixXd::Zero(dim_y, dim_y);
+  R(0, 0) = twist_stddev_vx_ * twist_stddev_vx_ * kf_dt_ * kf_dt_; // for vx
+  R(1, 1) = twist_stddev_wz_ * twist_stddev_wz_ * kf_dt_ * kf_dt_; // for wz
+
+  /* In order to avoid a large change by update, measurement update is performed by dividing at every step. */
+  R *= (kf_rate_ / twist_rate_);
+
+  kf_.updateWithDelay(y, C, R, delay_step);
+};
+
+/*
+ * mahalanobisGate
+ */
+bool KalmanFilterNode::mahalanobisGate(const double &dist_max, const Eigen::MatrixXd &x,
+                                       const Eigen::MatrixXd &obj_x, const Eigen::MatrixXd &cov)
+{
+  Eigen::MatrixXd mahalanobis_squared = (x - obj_x).transpose() * cov.inverse() * (x - obj_x);
+  DEBUG_INFO("measurement update: mahalanobis = %f, gate limit = %f", std::sqrt(mahalanobis_squared(0)), dist_max);
+  if (mahalanobis_squared(0) > dist_max * dist_max)
+  {
+    return false;
+  }
+
+  return true;
+}
+
+/*
+ * publishEstimatedPose
+ */
+void KalmanFilterNode::publishEstimatedPose()
+{
+  Eigen::MatrixXd X(dim_x_, 1);
+  Eigen::MatrixXd P(dim_x_, dim_x_);
+  kf_.getCurrentX(X);
+  kf_.getCurrentP(P);
+
+  /* publish latest pose */
+  pub_pose_.publish(current_kf_pose_);
+
+  /* publish latest pose with covariance */
+  geometry_msgs::PoseWithCovarianceStamped pose_cov;
+  pose_cov.header.stamp = ros::Time::now();
+  pose_cov.header.frame_id = current_kf_pose_.header.frame_id;
+  pose_cov.pose.pose = current_kf_pose_.pose;
+  for (int i = 0; i < 36; ++i)
+  {
+    pose_cov.pose.covariance[i] = 0.0;
+  }
+  pose_cov.pose.covariance[0] = P(IDX::X, IDX::X);      // x, x
+  pose_cov.pose.covariance[1] = P(IDX::X, IDX::Y);      // x, y
+  pose_cov.pose.covariance[5] = P(IDX::X, IDX::YAW);    // x, yaw
+  pose_cov.pose.covariance[6] = P(IDX::Y, IDX::X);      // y, x
+  pose_cov.pose.covariance[7] = P(IDX::Y, IDX::Y);      // y, y
+  pose_cov.pose.covariance[11] = P(IDX::Y, IDX::YAW);   // y, yaw
+  pose_cov.pose.covariance[30] = P(IDX::YAW, IDX::X);   // yaw, x
+  pose_cov.pose.covariance[31] = P(IDX::YAW, IDX::Y);   // yaw, y
+  pose_cov.pose.covariance[35] = P(IDX::YAW, IDX::YAW); // yaw, yaw
+  pub_pose_cov_.publish(pose_cov);
+
+  /* publish latest twist */
+  pub_twist_.publish(current_kf_twist_);
+
+  /* publish yaw bias */
+  std_msgs::Float64 yawb;
+  yawb.data = X(IDX::YAWB);
+  pub_yaw_bias_.publish(yawb);
+
+  /* debug my ndt */
+  if (current_ndt_pose_ptr_ != nullptr)
+  {
+    geometry_msgs::PoseStamped p;
+    p = *current_ndt_pose_ptr_;
+    p.header.stamp = ros::Time::now();
+    pub_ndt_pose_.publish(p);
+  }
+
+  /* debug publish */
+  double RAD2DEG = 180.0 / 3.141592;
```

Comment:
I think this code should use mathutil. (But not merged)
https://github.com/CPFL/Autoware/blob/ae86b06f7638b8b3b4a12381a2844d26ea569f3f/ros/src/common/libs/amathutils_lib/include/amathutils_lib/unit_convert.hpp

Reply:
(@sgermanserrano Because there is a lot of such code in autoware, I want to commonize)

Reply:
@yukkysaito that is already available [here](https://github.com/CPFL/Autoware/blob/e8fe86bf72e21bc84d52b4f198b6d8b2c3339f6f/ros/src/common/libs/amathutils_lib/include/amathutils_lib/amathutils.hpp#L18)

Reply:
@yukkysaito @sgermanserrano Thank you for your review and comments. I checked the current amathutils.hpp, but there is ```using namespace std;```  in there. I haven't checked if it works in this kalman code, but I think it is a little bit dangerous... Could I use my original code for this review?

Reply:
yes. that's right. 
I talked to the author who made it. He said that he wanted us to remake amathutils because he created it temporarily